modules/xfem/src/userobjects/MeshCut3DUserObject.C

```diff
@@ -0,0 +1,1106 @@
+    std::vector<Real> cut_pos;
+
+    for (unsigned int j = 0; j < n_edges; j++)
+    {
+      // This returns the lowest-order type of side.
+      std::unique_ptr<Elem> curr_edge = curr_side->side(j);
+      if (curr_edge->type() != EDGE2)
+        mooseError("In cutElementByGeometry face edge must be EDGE2, but type is: ",
+                   libMesh::Utility::enum_to_string(curr_edge->type()),
+                   " base element type is: ",
+                   libMesh::Utility::enum_to_string(elem->type()));
+      Node * node1 = curr_edge->get_node(0);
+      Node * node2 = curr_edge->get_node(1);
+
+      elem_it = elem_begin;
+      for (; elem_it != elem_end; ++elem_it)
+      {
+        std::vector<Point> vertices;
+        Elem * cut_elem = *elem_it;
+
+        for (unsigned int i = 0; i < cut_elem->n_nodes(); i++)
+        {
+          Node * node = cut_elem->get_node(i);
+          // Point * pt = dynamic_cast<Point *>(node);
+          Point & this_point = *node;
+          vertices.push_back(this_point);
+        }
+
+        Point intersection;
+        if (intersectWithEdge(*node1, *node2, vertices, intersection))
+        {
+          cut_edges.push_back(j);
+          cut_pos.push_back(getRelativePosition(*node1, *node2, intersection));
+        }
+      }
+    }
+
+    if (cut_edges.size() == 2)
+    {
+      elem_cut = true;
+      CutFace mycut;
+      mycut.face_id = i;
+      mycut.face_edge.push_back(cut_edges[0]);
+      mycut.face_edge.push_back(cut_edges[1]);
+      mycut.position.push_back(cut_pos[0]);
+      mycut.position.push_back(cut_pos[1]);
+      cut_faces.push_back(mycut);
+    }
+  }
+  return elem_cut;
+}
+
+bool
+MeshCut3DUserObject::cutFragmentByGeometry(std::vector<std::vector<Point>> & /*frag_edges*/,
+                                           std::vector<CutEdge> & /*cut_edges*/,
+                                           Real /*time*/) const
+{
+  mooseError("invalid method for 3D mesh cutting");
+  return false;
+}
+
+bool
+MeshCut3DUserObject::cutFragmentByGeometry(std::vector<std::vector<Point>> & /*frag_faces*/,
+                                           std::vector<CutFace> & /*cut_faces*/,
+                                           Real /*time*/) const
+{
+  // TODO: Need this for branching in 3D
+  mooseError("cutFragmentByGeometry not yet implemented for 3D mesh cutting");
+  return false;
+}
+
+bool
+MeshCut3DUserObject::intersectWithEdge(const Point & p1,
+                                       const Point & p2,
+                                       const std::vector<Point> & vertices,
+                                       Point & pint) const
+{
+  bool has_intersection = false;
+
+  Point normal;
+  Point center;
+  findElemNormal(vertices, normal, center);
+
+  double plane_point[3] = {center(0), center(1), center(2)};
+  double planenormal[3] = {normal(0), normal(1), normal(2)};
+  double edge_point1[3] = {p1(0), p1(1), p1(2)};
+  double edge_point2[3] = {p2(0), p2(1), p2(2)};
```

Comment:
In general, we avoid `double` in MOOSE. The exception is for pure utilities. Otherwise you should use `Real`.